### PR TITLE
CROSSDATA-33: function calls pushdown using spark plans

### PR DESCRIPTION
**Pull request by pfcoperez**
This pull-request enables native functions calls within non-purely-native Spark execution plans.

**A piece of warning:**

After merging it, Crossdata will require the spark-cassandra-connector to have the changes described by its [PR #864][1]

[1]:
https://github.com/datastax/spark-cassandra-connector/pull/864

### CODE REVIEW

File: cassandra/src/main/scala/org/apache/spark/sql/cassandra/CassandraXDSourceRelation.scala [hunk 1]
```diff
@@ -153,31 +175,59 @@ class CassandraXDSourceRelation(
...
+
       case sources.In(attribute, values) =>
         (quote(attribute) + " IN " + values.map(_ => "?").mkString("(", ", ", ")"), values.toSeq)
+
+      case sources.LessThan(attribute, f: AttributeReference) if(udfs contains f.toString) =>
+        udfvalcmp(attribute, "<", f)
       case sources.LessThan(attribute, value) => (s"${quote(attribute)} < ?", Seq(value))
+
+      case sources.LessThanOrEqual(attribute, f: AttributeReference) if(udfs contains f.toString) =>
+        udfvalcmp(attribute, "<", f)
```
Comment (jjlopezm) [inline cassandra/src/main/scala/org/apache/spark/sql/cassandra/CassandraXDSourceRelation.scala]: The symbol must be <=

File: cassandra/src/main/scala/org/apache/spark/sql/cassandra/CassandraXDSourceRelation.scala [hunk 2]
```diff
@@ -133,14 +135,34 @@ class CassandraXDSourceRelation(
...
-    val prunedRdd = maybeSelect(baseRdd, requiredColumns)
+    buildScan(requiredColumns, filters, Map.empty)
+  }
+
+  private def resolveUDFsReferences(strId: String, udfs: Map[String, NativeUDF]): Option[FunctionCallRef] =
+    udfs get(strId) map { udf =>
+      val actualParams = udf.children.collect {
+        case at: AttributeReference if(udfs contains at.toString) => Left(resolveUDFsReferences(at.toString, udfs).get)
+        case at: AttributeReference => Left(ColumnName(at.name))
+        case lit@Literal(_, DataTypes.StringType) => Right(s"$lit")
```
Comment (darroyocazorla) [inline cassandra/src/main/scala/org/apache/spark/sql/cassandra/CassandraXDSourceRelation.scala]: case lit@Literal(_, DataTypes.StringType) => Right(s"$lit") can be covered by case lit: Literal => Right(lit.toString)

**Comment by jjlopezm** [comment]
+1

**Comment by luismcl** [comment]
+1, por así decirlo...

**Pull request merged**